### ISSUE TITLE
Feature Request: Source handles expiring snapshots on the destination(s)

### ISSUE THREAD

**Issue by redmop** [jimsalterjrs/sanoid#106]
The source knows if the syncing is ok in znapzend, so it only expires snapshots on the source and destination only after all destinations are successfully synced. If Source syncs to DestA and DestB, and DestB is down, no snapshots are deleted on Source and DestB until B is caught up. This makes for happier off-sites through slow links.

It also protects the destination snapshots if the source is down.

**Comment by jimsalterjrs** [comment]
Sanoid already refuses to delete snapshots that would otherwise expire if it doesn't have *enough* of them on hand.  For example, if you have the policy `hourly=36` and you have 36 hourly snapshots on hand, but all of them are from nine days ago, the deletion policy _will not trigger_ - because `hourly=36` means both less than 36 hours old, _and_ at least 36 hourly snapshots on hand.  This protects you from issues with dropping WAN links as described.

Granted, older snapshots will still purge on the *source* - but that's what the syncoid snapshot is for; it keeps your place on source and target for as long as necessary, with the syncoid snapshot only being removed after a successful newer replication makes it obsolete.

Meanwhile, the more you tie the source and target together, the more opportunities you have for things to go wonky.  The only real benefit I can see here is "not having to run Sanoid on the remote machine" - but you *lose* an awful lot in that trade, and it's not like Sanoid is a big burden for installation.

Having Sanoid run on both ends also gives you the opportunity to have _separate_ policies on the local and remote sides; eg SSD storage in prod with fewer snapshots retained, rust storage in DR with lots and lots of snapshots retained.

This is a #wontfix unless you can come up with a more compelling argument.

**Comment by jimsalterjrs** [comment]
I'm assigning this to you @redmop since I'm blocking on waiting to see if you've got an argument that sways me. =)

If you got nothin', feel free to close it yourself, or I'll close it after a while (to leave some opportunity for discussion).

**Comment by redmop** [comment]
We can't use the syncoid snapshot though, with multiple destinations, unless we have a way to clean them up on the destination. Marker snapshots for DestB will be all over the place on DestA and will never be purged. The same problem exists on DestA with DestB snapshots.
We could use zfs send -i instead of -I if we loop through all the snapshots we are about individually, but that comes with it's own overhead.
If we use the --no-sync-snap, we lose the marker snapshots, and local expiry can delete something needed remotely.

What if we had a dataset property along the lines of syncedwith=DestA,DestB or SyncedwithDestA=true in each snapshot or perhaps lastsynceddest= property that moves with the latest synced snapshots?

**Comment by jimsalterjrs** [comment]
I like the dataset property idea.  I'm not really familiar with working with custom dataset properties though.

Can you give me some command-line examples of setting and reading a custom dataset property, and how you would envision syncoid setting them?  Perhaps along with a `sanoid.conf` mock directive that would instruct Sanoid to check for them before purging snapshots?

**Comment by redmop** [comment]
Give me a couple days. Znapzend does this, but it has been a while since I've toiled over that code.

**Comment by nirvdrum** [comment]
I'm looking at backing up my Ubuntu-based workstation running ZFS to my FreeNAS server. FreeNAS really doesn't want you messing with the base install and frankly I don't want to either, since it's apt to break on upgrades.

Unfortunately, you also can't really manage the host ZFS from within a jail. So, this is a use case where I'd really like the local machine to handle the expiring of snapshots on the remote machine. Essentially, I'd like to treat the remote machine as an appliance.

**Comment by phreaker0** [comment]
as already stated, this won't make it into sanoid
